src/documentalist.ts

```diff
@@ -36,49 +43,110 @@ const RESERVED_WORDS = [
-    documentFiles: (files: IFile[]) => IDocumentalistData;
+    /**
+     * Iterates over all plugins, passing all matching files to each in turn.
+     * The resulting output for each plugin is combined into the resulting
+     * documentation object.
+     *
+     * The documentation object has a composite type of all the plugin data
+     * types.
+     */
+    documentFiles: (files: IFile[]) => T;
 
     /**
-     * Render a block of content by extracting metadata front matter and
-     * splitting text content into rendered HTML strings and `{ tag, value }` objects.
+     * Render a block of content by extracting metadata (YAML front matter) and
+     * splitting text content into markdown-rendered HTML strings and `{ tag,
+     * value }` objects.
+     *
+     * To prevent special strings like "@include" from being parsed, a reserved
+     * tag words array may be provided, in which case the line will be left as
+     * is.
      */
     renderBlock: (blockContent: string, reservedTagWords?: string[]) => IBlock;
 
-    /** Use a plugin to process the files matching the pattern. */
-    use: (pattern: RegExp, plugin: IPlugin<any>) => IApi;
+    /**
+     * Adds the plugin to Documentalist. Returns a new instance of Documentalist
+     * with a template type that includes the data from the plugin. This way the
+     * `documentFiles` and `documentGlobs` methods will return an object that is
+     * already typed to include the plugin output.
+     *
+     * The Plugin is applied to all files whose absolute path matches the
+     * supplied pattern.
+     *
+     * @param pattern - A regexp pattern or a file extension string like "js"
+     * @param plugin - The plugin implementation
+     * @returns A new instance of `Documentalist` with an extended type
+     */
+    use: <P>(pattern: RegExp | string, plugin: IPlugin<P>) => IApi<T & P>;
+
+    /**
+     * Returns a new instance of Documentalist with no plugins.
+     */
+    clearPlugins(): IApi<void>
 }
 
-export interface IOptions {
-    markedOptions?: MarkedOptions;
+/**
+ * The output of `renderBlock` which parses a long form documentation block into
+ * metadata, rendered markdown, and tags.
+ */
+export interface IBlock {
+    /**
+     * The original string content block
+     */
+    content: string;
+
+    /**
+     * Parsed YAML front matter (if any) or {}.
+     */
+    metadata: any;
+
+    /**
+     * An array of markdown-rendered HTML or tags.
+     */
+    renderedContent: StringOrTag[];
 }
 
-export class Documentalist implements IApi {
-    private plugins: Array<{ pattern: RegExp, plugin: IPlugin<any> }> = [];
+/**
+ * Plugins are stored with the regex used to match against file paths.
+ */
+export interface IPluginEntry<T> {
+    pattern: RegExp;
+    plugin: IPlugin<T>;
+}
 
-    constructor(private markedOptions: MarkedOptions = {}) {
-        this.use(/\.md$/, new MarkdownPlugin())
+export class Documentalist<T> implements IApi<T> {
```

Comment:
Well yeah, but it would be basically `T extends any` because we don't really have any restrictions on the data that plugins can return other than they're an object.